x/evm/keeper/keeper.go

```diff
@@ -37,9 +42,10 @@ type Keeper struct {
 	bankKeeper    types.BankKeeper
 	stakingKeeper types.StakingKeeper
 
-	ctx sdk.Context
-	// set in `BeginCachedContext`, used by `GetCommittedState` api.
-	committedCtx sdk.Context
+	// initialized in `WithContext`,
+	// new cached context is pushed in `Snapshot`.
+	cachedContexts []cachedContext
```

Comment:
I think this part is relevant enough also to have a separate file. And you could only exposed 

` getContext() `
` Snapshot() ` 
` Commit() `
` Revert()`

That would make the keeper cleaner

Reply:
`Snapshot()` and `RevertToSnapshot()` are part of the `StateDB` interface, so I'd keep them there 👍

Reply:
yeah but if we decide to have a dedicated file `keeper/context.go` and move there all the context implementation logic using stack, we could go further in the abstraction.

 `Snapshot()` in StateDB could just call  a `Snapshot()` of the the keeperContext.

Reply:
thanks, the new api is almost the same as suggested.